text_to_image/backend_pytorch.py

```diff
@@ -0,0 +1,415 @@
+from typing import Optional, List, Union
+import os
+import torch
+import logging
+import backend
+from diffusers import (
+    StableDiffusionXLPipeline,
+    StableDiffusionXLImg2ImgPipeline,
+)
+from diffusers import EulerDiscreteScheduler
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger("backend-pytorch")
+
+
+class BackendPytorch(backend.Backend):
+    def __init__(
+        self,
+        model_path=None,
+        model_id="xl",
+        guidance=8,
+        steps=20,
+        device="cuda",
+        precision="fp32",
+        negative_prompt="normal quality, low quality, worst quality, low res, blurry, nsfw, nude",
+    ):
+        super(BackendPytorch, self).__init__()
+        self.model_path = model_path
+        if model_id == "xl":
+            self.model_id = "stabilityai/stable-diffusion-xl-base-1.0"
+        else:
+            raise ValueError(f"{model_id} is not a valid model id")
+
+        self.device = device if torch.cuda.is_available() else "cpu"
+        if precision == "fp16":
+            self.dtype = torch.float16
+        elif precision == "bf16":
+            self.dtype = torch.bfloat16
+        else:
+            self.dtype = torch.float32
+
+        if torch.cuda.is_available():
+            self.local_rank = 0
+            self.world_size = 1
+
+        self.guidance = guidance
+        self.steps = steps
+        self.generator_seed = 42
+        self.negative_prompt = negative_prompt
+        self.max_length_neg_prompt = 77
+        self.max_length_refiner_neg_prompt = 77
+
+    def version(self):
+        return torch.__version__
+
+    def name(self):
+        return "pytorch-SUT"
+
+    def image_format(self):
+        return "NCHW"
+
+    def load(self):
+        if self.model_path is None:
+            log.warning(
+                "Model path not provided, running with default hugging face weights\n"
+                "This may not be valid for official submissions"
+            )
+            self.scheduler = EulerDiscreteScheduler.from_pretrained(
+                self.model_id, subfolder="scheduler"
+            )
+            self.pipe = StableDiffusionXLPipeline.from_pretrained(
+                self.model_id,
+                scheduler=self.scheduler,
+                safety_checker=None,
+                add_watermarker=False,
+                variant="fp16" if (self.dtype == torch.float16) else None,
+                torch_dtype=self.dtype,
+            )
+            # self.pipe.unet = torch.compile(self.pipe.unet, mode="reduce-overhead", fullgraph=True)
+            self.refiner_pipe = StableDiffusionXLImg2ImgPipeline.from_pretrained(
```

Comment:
If possible can we remove code related to refiner network to keep the code simple?

Reply:
Removed